src/opera/instance/topology.py

```diff
@@ -11,33 +11,27 @@ def __init__(self, storage, nodes):
             node.instantiate_relationships()
             node.read()
 
-    def deploy(self):
-        # TODO(@tadeboro): This is where parallelization magic should go.
-        # Currently, we are running a really stupid O(n^3) algorithm, but
-        # unless we get to the templates with millions of node instances, we
-        # should be fine.
-
-        do_deploy = True
-        while do_deploy:
-            do_deploy = False
-            for node in self.nodes.values():
-                if not node.deployed and node.ready_for_deploy:
-                    node.deploy()
-                    do_deploy = True
-
-    def undeploy(self):
-        # TODO(@tadeboro): This is where parallelization magic should go.
-        # Currently, we are running a really stupid O(n^3) algorithm, but
-        # unless we get to the templates with millions of node instances, we
-        # should be fine.
-
-        do_undeploy = True
-        while do_undeploy:
-            do_undeploy = False
-            for node in self.nodes.values():
-                if not node.undeployed and node.ready_for_undeploy:
-                    node.undeploy()
-                    do_undeploy = True
+    def deploy(self, num_workers=None):
+        with NodeExecutor(num_workers) as executor:
+            do_deploy = True
+            while do_deploy:
+                for node in self.nodes.values():
+                    if (not node.deployed
+                            and node.ready_for_deploy
+                            and executor.not_submitted(node.tosca_id)):
+                        executor.submit_operation(node.deploy, node.tosca_id)
+                do_deploy = executor.wait_results()
+
+    def undeploy(self, num_workers=None):
```

Comment:
I've tested your changes locally on the example found in `tests/integration/misc-tosca-types` by running deploy and then undeploy, both with two workers and got the following output:

```console
(.venv) anzoman@ubuntu:~/Desktop/xopera-opera$ opera deploy -i tests/integration/misc-tosca-types/inputs.yaml -w 2 tests/integration/misc-tosca-types/service-template.yaml
[Worker_0]   Deploying my-workstation1_0
[Worker_1]   Deploying my-workstation2_0
[Worker_1]   Deployment of my-workstation2_0 complete
[Worker_1]   Deploying file_0
[Worker_1]     Executing create on file_0
[Worker_0]   Deployment of my-workstation1_0 complete
[Worker_0]   Deploying noimpl_0
[Worker_0]   Deployment of noimpl_0 complete
[Worker_0]   Deploying hello_0
[Worker_0]     Executing create on hello_0
[Worker_0]   Deployment of hello_0 complete
[Worker_0]   Deploying test_0
[Worker_0]     Executing create on test_0
[Worker_1]   Deployment of file_0 complete
[Worker_1]   Deploying setter_0
[Worker_1]     Executing create on setter_0
[Worker_0]   Deployment of test_0 complete
[Worker_1]   Deployment of setter_0 complete
(.venv) anzoman@ubuntu:~/Desktop/xopera-opera$ opera undeploy -w 2
[Worker_0]   Undeploying my-workstation2_0
[Worker_1]   Undeploying file_0
[Worker_1]     Executing delete on file_0
[Worker_0]   Undeployment of my-workstation2_0 complete
[Worker_0]   Undeploying noimpl_0
[Worker_0]   Undeployment of noimpl_0 complete
[Worker_0]   Undeploying setter_0
[Worker_0]   Undeployment of setter_0 complete
[Worker_0]   Undeploying test_0
[Worker_0]   Undeployment of test_0 complete
[Worker_0]   Undeploying hello_0
[Worker_0]   Undeployment of hello_0 complete
[Worker_0]   Undeploying my-workstation1_0
[Worker_0]   Undeployment of my-workstation1_0 complete
[Worker_1]   Undeployment of file_0 complete
```

And I'm just wondering if it is right that for the deploy part operations were more distributed for the operations (e.g `Worker_1` was used for deploying `my-workstation2_0`, `file_0` and `setter_0` nodes) and for undeploy the situation different (like `Worker_1` was used just for undeploying the `file_0` node). Is this ok because the undeployement graph is different because it is created  starting from the opposite side?

Reply:
I suggest to keep this part of Tadej's comment:

```
        # Currently, we are running a really stupid O(n^3) algorithm, but
        # unless we get to the templates with millions of node instances, we
        # should be fine.
```

Reply:
Well in this case it basically happens because `test_0` node has `create `operation implemented but does not have `delete `operation. So there is no need for additional thread as `test_0` undeploy exits instantly.